packages/editor/src/components/block-switcher/index.js

```diff
@@ -131,14 +139,12 @@ export class BlockSwitcher extends Component {
 }
 
 export default compose(
-	withSelect( ( select, ownProps ) => {
-		const { getBlock, getBlockRootClientId, getTemplateLock } = select( 'core/editor' );
+	withSelect( ( select, { clientIds } ) => {
+		const { getBlocksByClientId, getBlockRootClientId, getInserterItems } = select( 'core/editor' );
+		const rootClientId = getBlockRootClientId( first( castArray( clientIds ) ) );
 		return {
-			blocks: ownProps.clientIds.map( getBlock ),
-			isLocked: some(
-				castArray( ownProps.clientIds ),
-				( clientId ) => !! getTemplateLock( getBlockRootClientId( clientId ) )
-			),
+			blocks: getBlocksByClientId( clientIds ),
+			inserterItems: getInserterItems( rootClientId ),
```

Comment:
Have you looked into using `canInsertBlockType` instead of `getInserterItems`? The latter takes reusable blocks into account which is unnecessary in this case because you can't transform a block into an existing reusable block. It's better that we do as little work as possible to keep the UI feeling responsive.

Reply:
I have thought about this and it seemed getInserterItems was the best option. 
We don't want just to know if a block can be inserted we want to know all the blocks that can be inserted so we can render the UI. We can not do a simple filter using canInsertBlockType inline because that would create a new array reference each time triggering unnecessary rerenders. So the alternative would be to create a new selector that just returns all the allowed blocks without the shared blocks. Besides the available blocks, we also need the frecency ( to sort the items ) and getInserterItems already provides it.
getInserterItems is cached and used in the inserter which is rerendered each time a block changes so I think each time we use the selector here we are just retrieving the value from the cache. If we had a specific selector we would not be able to share the cache.

Reply:
Makes sense 👍 – thanks for explaining!